### PR TITLE
Fix two bugs of experiment resume and improve unittest of nnimanager

### PR DESCRIPTION
**Pull request by QuanluZhang**
### Description ###
- [x] fix the two bugs induced by the feature that the interrupted trials are resumed when an experiment is resumed:
  - [x] trial number is counted incorrectly when an experiment is resumed and some interrupted trials are rerun.
  - [x] fix the bug that adding customized trial makes tuner/msg_dispatcher crash
- [x] grealy improve the unit tests of nnimanager.

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/runtime/msg_dispatcher.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ def save_checkpoint(self):
     def handle_initialize(self, data):
         """Data is search space
         """
+        _logger.info('Initial search space: %s, %s', type(data), data)
```
Comment (liuzhe-lz) [inline nni/runtime/msg_dispatcher.py:90]: If the level is INFO, it should not contain `type(data)`.

File: ts/nni_manager/common/globals/unittest.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ import path from 'path';
 
 import type { NniManagerArgs } from './arguments';
 import type { LogStream } from './log_stream';
+import { initLogStream } from './log_stream';
```
Comment (liuzhe-lz) [inline ts/nni_manager/common/globals/unittest.ts:27]: For style consistency, please merge these lines to `import { LogStream, initLogStream } from ...`.

File: ts/nni_manager/common/globals/unittest.ts [hunk 3]
```diff
@@ -60,11 +61,13 @@ export function resetGlobals(): void {
...
-        writeLineSync: (_line: string): void => { /* dummy */ },
-        close: async (): Promise<void> => { /* dummy */ }
-    };
+    // const logStream = {
+    //     writeLine: (_line: string): void => { /* dummy */ },
+    //     writeLineSync: (_line: string): void => { /* dummy */ },
+    //     close: async (): Promise<void> => { /* dummy */ }
+    // };
+    // modify logStream in unittest globals to create nnimanager.log for easy debugbility
+    const logStream = initLogStream(args, paths);
```
Comment (liuzhe-lz) [inline ts/nni_manager/common/globals/unittest.ts:70]: It needs to close the previous stream.

File: ts/nni_manager/core/tuner_command_channel/websocket_channel.ts [hunk 4]
```diff
@@ -33,6 +33,8 @@ export interface WebSocketChannel {
  *  Remember to invoke ``await channel.init()`` before doing anything else.
  **/
 export function getWebSocketChannel(): WebSocketChannel {
+    // return channelSingleton;
+    channelSingleton = new WebSocketChannelImpl();
```
Comment (liuzhe-lz) [inline ts/nni_manager/core/tuner_command_channel/websocket_channel.ts:37]: Update the docstr.

File: ts/nni_manager/package.json [hunk 5]
```diff
@@ -4,11 +4,12 @@
   "license": "MIT",
   "scripts": {
     "build": "tsc",
-    "test": "nyc --reporter=cobertura --reporter=text mocha test/**/*.test.ts",
+    "test": "nyc --reporter=cobertura --reporter=text mocha test/**/nnimanager.test.ts",
```
Comment (liuzhe-lz) [inline ts/nni_manager/package.json:7]: ?

File: ts/nni_manager/package.json [hunk 6]
```diff
@@ -4,11 +4,12 @@
...
     "mocha": "mocha",
     "eslint": "eslint . --ext .ts"
   },
   "dependencies": {
+    "@types/mocha": "^9.0.0",
```
Comment (liuzhe-lz) [inline ts/nni_manager/package.json:12]: This is dev dependency.

File: ts/nni_manager/package.json [hunk 7]
```diff
@@ -21,6 +22,7 @@
     "js-base64": "^3.7.2",
     "kubernetes-client": "^6.12.1",
     "lockfile": "^1.0.4",
+    "mocha": "^9.0.0",
```
Comment (liuzhe-lz) [inline ts/nni_manager/package.json:25]: This is also dev dependency. And why it's downgraded?

File: nni/runtime/msg_dispatcher.py (hunk 1 above)
Comment (QuanluZhang) [reply nni/runtime/msg_dispatcher.py:90]: sure, removed

File: ts/nni_manager/common/globals/unittest.ts (hunk 2 above)
Comment (QuanluZhang) [reply ts/nni_manager/common/globals/unittest.ts:27]: updated

File: ts/nni_manager/common/globals/unittest.ts (hunk 3 above)
Comment (QuanluZhang) [reply ts/nni_manager/common/globals/unittest.ts:70]: updated

File: nni/recoverable.py [hunk 8]
```diff
@@ -43,7 +43,9 @@ def recover_parameter_id(self, data) -> int:
         self.recovered_max_param_id = previous_max_param_id
         return previous_max_param_id
 
-    def is_created_in_previous_exp(self, param_id: int) -> bool:
+    def is_created_in_previous_exp(self, param_id: int|None) -> bool:
```
Comment (liuzhe-lz) [inline nni/recoverable.py:46]: space around `|`

File: ts/nni_manager/core/tuner_command_channel/websocket_channel.ts [hunk 9]
```diff
@@ -190,7 +190,13 @@ class WebSocketChannelImpl implements WebSocketChannel {
     }
 }
 
-const channelSingleton: WebSocketChannelImpl = new WebSocketChannelImpl();
+let channelSingleton: WebSocketChannelImpl = new WebSocketChannelImpl();
+
+// NOTE: this function is only for unittest of nnimanager,
+// because resuming an experiment should reset websocket channel.
+export function resetChannelSingleton(): void {
```
Comment (liuzhe-lz) [inline ts/nni_manager/core/tuner_command_channel/websocket_channel.ts:197]: Move it to UnitTestHelpers

File: ts/nni_manager/package-lock.json [hunk 10]
```diff
@@ -9,6 +9,7 @@
       "version": "999.0.0-developing",
       "license": "MIT",
       "dependencies": {
+        "@types/mocha": "^9.0.0",
```
Comment (liuzhe-lz) [inline ts/nni_manager/package-lock.json:12]: These are dev dependencies

File: ts/nni_manager/common/globals/unittest.ts [hunk 11]
```diff
@@ -66,6 +66,11 @@ export function resetGlobals(): void {
         writeLineSync: (_line: string): void => { /* dummy */ },
         close: async (): Promise<void> => { /* dummy */ }
     };
+    // // init log stream here to create nnimanager.log for easy debugbility
```
Comment (liuzhe-lz) [inline ts/nni_manager/common/globals/unittest.ts:69]: Remove debug stuff

File: ts/nni_manager/common/globals/unittest.ts [hunk 12]
```diff
@@ -23,7 +23,7 @@ import os from 'os';
 import path from 'path';
 
 import type { NniManagerArgs } from './arguments';
-import type { LogStream } from './log_stream';
+import { LogStream, initLogStream } from './log_stream';
```
Comment (liuzhe-lz) [inline ts/nni_manager/common/globals/unittest.ts:26]: not used

File: ts/nni_manager/core/nniDataStore.ts [hunk 13]
```diff
@@ -67,7 +67,7 @@ class NNIDataStore implements DataStore {
         return this.db.queryLatestExperimentProfile(experimentId);
     }
 
-    public storeTrialJobEvent(
+    public async storeTrialJobEvent(
```
Comment (liuzhe-lz) [inline ts/nni_manager/core/nniDataStore.ts:70]: If it's made async, try-catch should be preferred over promise.catch()

Comment (QuanluZhang) [reply ts/nni_manager/core/nniDataStore.ts:70]: let me remove `async`.

**Review by liuzhe-lz — APPROVED** [review]

**Review by J-shang — APPROVED** [review]

**Pull request merged**